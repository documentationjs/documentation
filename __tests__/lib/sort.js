var sort = require('../../src/sort'),
  path = require('path');

test('sort stream alphanumeric', function() {
  var apples = { context: { sortKey: 'a' }, name: 'apples' };
  var carrot = { context: { sortKey: 'b' }, name: 'carrot' };
  var banana = { context: { sortKey: 'c' }, name: 'bananas' };

  expect(sort([apples, carrot, banana])).toEqual([apples, carrot, banana]);

  expect(sort([carrot, apples, banana])).toEqual([apples, carrot, banana]);
});

test('sort stream with configuration', function() {
  var apples = { context: { sortKey: 'a' }, name: 'apples' };
  var carrot = { context: { sortKey: 'b' }, name: 'carrot' };
  var bananas = { context: { sortKey: 'c' }, name: 'bananas' };

  expect(
    sort([apples, carrot, bananas], {
      toc: ['carrot', 'bananas']
    })
  ).toEqual([carrot, bananas, apples]);
});

test('sort stream with configuration and a section', function() {
  var apples = { context: { sortKey: 'a' }, name: 'apples' };
  var carrot = { context: { sortKey: 'b' }, name: 'carrot' };
  var bananas = { context: { sortKey: 'c' }, name: 'bananas' };

  var section = {
    name: 'This is the banana type',
    description: 'here lies bananas'
  };

  var sectionMarkdown = {
    name: 'This is the banana type',
    description: {
      type: 'root',
      children: [
        {
          type: 'paragraph',
          children: [
            {
              type: 'text',
              value: 'here lies bananas',
              position: {
                start: {
                  line: 1,
                  column: 1,
                  offset: 0
                },
                end: {
                  line: 1,
                  column: 18,
                  offset: 17
                },
                indent: []
              }
            }
          ],
          position: {
            start: {
              line: 1,
              column: 1,
              offset: 0
            },
            end: {
              line: 1,
              column: 18,
              offset: 17
            },
            indent: []
          }
        }
      ],
      position: {
        start: {
          line: 1,
          column: 1,
          offset: 0
        },
        end: {
          line: 1,
          column: 18,
          offset: 17
        }
      }
    },
    kind: 'note',
    path: [
      {
        name: 'This is the banana type',
        scope: 'static'
      }
    ]
  };

  expect(
    sort([apples, carrot, bananas], {
      toc: ['carrot', section, 'bananas']
    })
  ).toEqual([carrot, sectionMarkdown, bananas, apples]);
});

test('sort an already-sorted stream containing a section/description', function() {
  // this happens in the 'serve' task
  var apples = { context: { sortKey: 'a' }, name: 'apples' };
  var carrot = { context: { sortKey: 'b' }, name: 'carrot' };
  var bananas = { context: { sortKey: 'c' }, name: 'bananas' };

  var section = {
    name: 'This is the banana type',
    description: 'here lies bananas'
  };
  var sectionMarkdown = {
    name: 'This is the banana type',
    description: {
      type: 'root',
      children: [
        {
          type: 'paragraph',
          children: [
            {
              type: 'text',
              value: 'here lies bananas',
              position: {
                start: {
                  line: 1,
                  column: 1,
                  offset: 0
                },
                end: {
                  line: 1,
                  column: 18,
                  offset: 17
                },
                indent: []
              }
            }
          ],
          position: {
            start: {
              line: 1,
              column: 1,
              offset: 0
            },
            end: {
              line: 1,
              column: 18,
              offset: 17
            },
            indent: []
          }
        }
      ],
      position: {
        start: {
          line: 1,
          column: 1,
          offset: 0
        },
        end: {
          line: 1,
          column: 18,
          offset: 17
        }
      }
    },
    kind: 'note',
    path: [
      {
        name: 'This is the banana type',
        scope: 'static'
      }
    ]
  };

  var config = {
    toc: ['carrot', section, 'bananas']
  };

  var sortOnce = sort([apples, carrot, bananas], config);
  var sortTwice = sort(sortOnce, config);
  expect(sortTwice).toEqual([carrot, sectionMarkdown, bananas, apples]);
});

test('sort toc with files', function() {
  var apples = { context: { sortKey: 'a' }, name: 'apples' };
  var carrot = { context: { sortKey: 'b' }, name: 'carrot' };
  var bananas = { context: { sortKey: 'c' }, name: 'bananas' };

  var snowflake = {
    name: 'snowflake',
    file: path.join(__dirname, '../fixture/snowflake.md')
  };

  expect(
    sort([apples, carrot, bananas], {
      toc: [snowflake]
    })
  ).toMatchSnapshot();
});

test('sort toc with files absolute path', function() {
  var apples = { context: { sortKey: 'a' }, name: 'apples' };
  var carrot = { context: { sortKey: 'b' }, name: 'carrot' };
  var bananas = { context: { sortKey: 'c' }, name: 'bananas' };

  var snowflake = {
    name: 'snowflake',
    file: path.join(__dirname, '../fixture/snowflake.md')
  };
  expect(
    sort([apples, carrot, bananas], {
      toc: [snowflake]
    })
  ).toMatchSnapshot();
});

test('sort toc with files absolute path', function() {
  var apples = {
    context: { sortKey: 'a' },
    name: 'apples',
    memberof: 'classB'
  };
  var carrot = {
    context: { sortKey: 'b' },
    name: 'carrot',
    memberof: 'classB'
  };
  var bananas = {
    context: { sortKey: 'c' },
    name: 'bananas',
    memberof: 'classB'
  };

  var snowflake = {
    name: 'snowflake',
    file: path.join(__dirname, '../fixture/snowflake.md')
  };
  expect(
    sort([carrot, apples, bananas], {
      sortOrder: 'alpha'
    })
  ).toMatchSnapshot();
});
