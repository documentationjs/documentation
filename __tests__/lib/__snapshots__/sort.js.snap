// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sort toc with files 1`] = `
Array [
  Object {
    "description": Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "position": Position {
                "end": Object {
                  "column": 16,
                  "line": 1,
                  "offset": 15,
                },
                "indent": Array [],
                "start": Object {
                  "column": 3,
                  "line": 1,
                  "offset": 2,
                },
              },
              "type": "text",
              "value": "The Snowflake",
            },
          ],
          "depth": 1,
          "position": Position {
            "end": Object {
              "column": 16,
              "line": 1,
              "offset": 15,
            },
            "indent": Array [],
            "start": Object {
              "column": 1,
              "line": 1,
              "offset": 0,
            },
          },
          "type": "heading",
        },
      ],
      "position": Object {
        "end": Object {
          "column": 1,
          "line": 2,
          "offset": 16,
        },
        "start": Object {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "root",
    },
    "kind": "note",
    "name": "snowflake",
    "path": Array [
      Object {
        "name": "snowflake",
        "scope": "static",
      },
    ],
  },
  Object {
    "context": Object {
      "sortKey": "a",
    },
    "name": "apples",
  },
  Object {
    "context": Object {
      "sortKey": "b",
    },
    "name": "carrot",
  },
  Object {
    "context": Object {
      "sortKey": "c",
    },
    "name": "bananas",
  },
]
`;

exports[`sort toc with files absolute path 1`] = `
Array [
  Object {
    "description": Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "position": Position {
                "end": Object {
                  "column": 16,
                  "line": 1,
                  "offset": 15,
                },
                "indent": Array [],
                "start": Object {
                  "column": 3,
                  "line": 1,
                  "offset": 2,
                },
              },
              "type": "text",
              "value": "The Snowflake",
            },
          ],
          "depth": 1,
          "position": Position {
            "end": Object {
              "column": 16,
              "line": 1,
              "offset": 15,
            },
            "indent": Array [],
            "start": Object {
              "column": 1,
              "line": 1,
              "offset": 0,
            },
          },
          "type": "heading",
        },
      ],
      "position": Object {
        "end": Object {
          "column": 1,
          "line": 2,
          "offset": 16,
        },
        "start": Object {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "root",
    },
    "kind": "note",
    "name": "snowflake",
    "path": Array [
      Object {
        "name": "snowflake",
        "scope": "static",
      },
    ],
  },
  Object {
    "context": Object {
      "sortKey": "a",
    },
    "name": "apples",
  },
  Object {
    "context": Object {
      "sortKey": "b",
    },
    "name": "carrot",
  },
  Object {
    "context": Object {
      "sortKey": "c",
    },
    "name": "bananas",
  },
]
`;

exports[`sort toc with files absolute path 2`] = `
Array [
  Object {
    "context": Object {
      "sortKey": "a",
    },
    "memberof": "classB",
    "name": "apples",
  },
  Object {
    "context": Object {
      "sortKey": "c",
    },
    "memberof": "classB",
    "name": "bananas",
  },
  Object {
    "context": Object {
      "sortKey": "b",
    },
    "memberof": "classB",
    "name": "carrot",
  },
]
`;
